packages/loader/container-loader/src/container.ts

```diff
@@ -674,17 +675,65 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
         this.resumeInternal();
     }
 
+    /**
+     * Waits until container connects to delta storage and gets up-to-date
+     * Useful when resolving URIs and hitting 404, due to container being loaded from (stale) snapshot and not being
+     * up to date. Host may chose to wait in such case and retry resolving URI.
+     * @returns true: container is up to date, it processed all the ops that were know at the time of first connection
+     *          false: storage does not provide indication of how far the client is.
+     */
+    public async waitCatchUp() {
+        return new Promise<boolean>((accept) => {
```

Comment:
i think you can remove this outer promise, and use use return/throw below rather than accept+return

Reply:
The problem is that accept callback is called within delta manager callback, so I need some kind of deferral. There is only one path where we can return right away, all others require callback instantiating and waiting for right thing to happen